### ISSUE TITLE
Chapter 8 - IO.of usage, example fix

### ISSUE THREAD

**Issue by zfeher** [MostlyAdequate/mostly-adequate-guide#342]
Hi' First of all the book is awesome so far hope you will continue someday. Currently I am going through the chapters and playing with the inline examples. 

1, In chapter 8 you define the `IO.of` method but you are using `new IO(...)` in the inline examples instead which first confused me a little bit, felt like we are doing double function wrapping (in my head I wanted to see `IO.of` instead of `new IO(...)` hence was the double wrap feel). 

Then I started replacing `new IO(...)` with `IO.of`:
```JavaSsript
IO.of(window)
IO.of(() => document.querySelectorAll(selector))
IO.of(window.location.href)
```
This version is better for the eyes and easier on the mind. However the second one didn't work out of the box. The wrapped function needed to be called (`map(f => f())`) before the other mappings which felt like an unnecessary overhead for me. So I modified the IO functor to not wrap functions:

```JavaScript
IO.of = function(x) {
  return new IO(
    typeof x === 'function' ? x : function() { return x; });
};
```

2, I've found the `findParam` function incomplete:
```JavaScript
//  findParam :: String -> IO Maybe [String]
var findParam = function(key) {
  return map(compose(Maybe.of, filter(compose(eq(key), head)), params), url);
};
```
Here filter either returns an array with one element (1 pair) or an empty array. Wrapping this in a Maybe felt unnecessary at first because you won't get a `null` or `undefined`. Plus you will still have the pair wrapped in an extra array. So I added an extra `head` after the filtering to remove the extra array wrap. This way it clearer why we are wrapping in a `Maybe`.

```JavaScript
//  findParam :: String -> IO Maybe [String]
var findParam = function(key) {
  return map(compose(Maybe.of, head, filter(compose(eq(key), head)), params), url);
};
```

**Comment by KtorZ** [comment]
Hey!
Glad that you appreciate the book.

1.
Your first remark is legit though you're going in a wrong direction ^.^
First, you have to remember one thing: we use the IO monad to represent effectful action - things that happen (or will happen) and that will return something when executed. We don't want to execute actions directly as it will unleash the effects, so we want to keep them safe in a container and be able to run them whenever we want.

In JavaScript, functions are first-class citizens so we can assign them to variable and refer to them without actually execute them. And this is why the `new IO(f)` constructor is for. When we have a function (something like `* -> *`), we wrap it directly. From this, we also derive an implicit implementation contract telling us that what's inside IO should always be a function.

However, there are cases where we also want to wrap pure value in an IO monad (to use them in an applicative way for instance). Therefore, we define the `IO.of` constructor to give us an ability to put those pure value inside an IO. How? By creating a function under the hood (in a way, an effect) which will return this pure value when ran. 
In this sense, `IO.of` is just a shortcut when we don't want to manually lift pure elements into the monad and you're right when you say that the very first example given could be

```
IO.of(window)
```

instead of

```
var io_window = new IO(function() {
  return window;
});
```
 
Yet, we can't modify the `IO.of` constructor the way you did as it completely changes the function signature (this is a bad JavaScript habit ^.^). 

```hs
-- Original version
IO.of :: a -> IO a

-- Modified version
IO.of :: a -> IO a -- When a is not a function
IO.of :: a -> a -- When a is a function
```

Do you see the issue here?  


2.
The `Maybe` looks indeed unnecessary here. I would simply remove it instead of adding a `head` in the composition since you may actually have more than one pair (e.g. `?key=value&key=value2`). 
However, I think the `Maybe` is there simply to emphasize on the fact that we could have a stack of monads (`IO(Monad([a]))`). A better example which also shows this behavior may be a better idea than simply removing the Maybe :)

**Comment by zfeher** [comment]
Thank you for your reply. Until now I didn't have the time to reply but finally I have some.

Regarding point 1, I think I understand the things you wrote but I can't see the signature change in my version of `IO.of` (both returns an `IO`) so I am going back to start.

There was the following example:
```JavaScript
new IO(() => document.querySelectorAll(selector))
```

I wanted to express this via `IO.of` as well like the other simple value cases (eg: `IO.of(window)`).

Simply writing the following resulted one more nesting then needed
```JavaScript
IO.of(() => document.querySelectorAll(selector))

// ==
new IO(
  () => 
    () => document.querySelectorAll(selector)
)

// not the same as the original example
new IO(() => document.querySelectorAll(selector))
```

So I've got the idea to update the `IO.of` constructor function to not wrap functions one more time. 

The modified `IO.of` works for (already) delayed functions like in the example but now I realized it will probably cause some problems if I want to pass function as a value to `IO.of`. And this may change the signature of `IO.of`. Can it be the following or still yours is the correct ?
```JavaScript
IO.of :: a -> IO a -- When a is not a function
IO.of :: (()->a) -> IO a -- When a is a function
```

So in the end I can't really modifiy `IO.of` :) Instead I need to use the `new IO(() => { ... })` in "special" cases like the above one.

**Comment by KtorZ** [comment]
Hey!

I'll go through several points you raised and try to explain them briefly :)

> I wanted to express this via IO.of as well like the other simple value cases (eg: IO.of(window)).

This "doesn't make much sense". You can think of `.of` as a mechanism to put a pure value (free of any side-effects) into your functor. This value is also called a *default minimal context* (cf [Chapter 9](https://drboolean.gitbooks.io/mostly-adequate-guide/content/ch9.html)) and is especially useful when you want to inject values which don't come from any side-effect in a chain of computation. In case of the `IO` monad, we are talking about Input/Output effects (which can include many things like writing to a file or *reading a DOM element*).

In plain English, you could "read" `IO.of(a)` as "put this value `a` inside an IO monad as if it was the result of an IO effect". In this implementation, we model `IO` effects by wrapping the surrounding call into a function. Hence, in that case, `IO.of(a)` is equivalent to `() => a`.

Now, by doing `IO.of(() => document.querySelectorAll(selector))`, you'd be asking to "create" an `IO` effect which returns back a function which returns back a node element. Another way to see this, considering the remark above, is to think of `() => document.querySelectorAll(selector)` as an `IO` and since there is no constraint about the kind of value that can be placed in a Monad, you're free to put an `IO` Monad inside another `IO` Monad (which what happen by doing `IO.of(...)`).

So, in essence `.of` and `new` are rather different concepts. Actually, `.of` is an important abstraction which is common to many `Monad` / `Functor` / `Applicative` whereas `new IO` is *our way* to cope with vanilla JavaScript not inherently having this concept of Monad. In the end, they have a different meaning. 

### TL; DR;

- `IO.of` lifts a pure value into the Monad. In return, you get a Monad which will behave as if the value was actually the result of a side-effect. This interface is the same and has the same meaning for every Monad.

- `new IO` constructs an IO Monad from a function. This interface is specific to the `IO` monad and its implementation. This enables turning vanilla effectful JavaScript functions to a controlled structure describing those effects (a.k.a the `IO` Monad)
